### ISSUE TITLE
Access material properties from GenericSection instance

### ISSUE THREAD

**Issue by DanielGMorenaFhecor** [fib-international/structuralcodes#135]
We cannot access the properties of the materials within an instance of `GenericSection`. 

Ideally, it should be possible to access these material properties to obtain further information without relying on previous instances of material classes.

**Comment by talledodiego** [comment]
By choice we decided to pass only constitutive laws that are needed for computing the response of the section. We did this choice in the beginning of the design when not all materials were already developed (in particular there was not reinforcement yet). In this way the user could pass a Material (and from there we take the ConstitutiveLaw) or directly a custom ConstitutiveLaw.

We can pass a Material if we think that there are needed properties in the material for the computation of sectional properties, but we should prevent the user to pass a custom ConstitutiveLaw and permit to pass only Materials in order to be sure that those properties are always available.

Can we discuss here more in detail what properties should be needed from the material at the section computation level?

**Comment by DanielGMorenaFhecor** [comment]
I see your point, and it's a good one. Separating concerns between the material and section classes is an important consideration. Our main concern, however, was tracking which material is used in each section, along with its specific parameters, because it's common to instantiate multiple materials with their respective sections. We recognize that assigning a name to a section in its constructor can help mitigate this issue to some extent as it improves trackability.

Personally (and I might be biased here, so take this with a grain of salt), I feel that a cross-section inherently involves a material (or multiple materials), so reducing the material to just a constitutive equation seems a bit off to me.

This discussion might also relate to Issue #147. One of my questions now is about the future direction of the cross-section class. For example, if we plan to implement a method for computing crack width (but it could be any other useful property) in a concrete section, like `section.wk(M,...)`, will that method be included in the `Section` class (or a derived `ConcreteSection` class)? Or should we keep such functionality separate in the `codes` module, a wrapper class, or a future `Element` superclass?

If we decide to include crack width calculations in the Section class (or an inherited ConcreteSection), it might be useful to also retain the material information for these calculations.

What are your thoughts on this?

**Comment by talledodiego** [comment]
I see your points.

When dealing with common section behavior (strenght, domain, moment-curvature) the constituve law is enough. I would not care too much about "trackability" in a backend: it is a issue more related to a front-end. We are creating a back-end where it is easier to re-run a cell than to do many checks on trackability. I feel that those aspects should be dealed by a front end. Anyhow you are right: names helps somehow this trackability.

On the other hand, I am interested in aspects of computation that may need the material class, rather than just the constitutive law. We should discuss here if we need some stuff for computations. 

In a certain way crack width is related to an Element class made as a compound of one or more sections. 
On the other hand I also see that, as often happens, some "structural level" characteristics are seen by most engineers as "sectional level" like for instance crack width computation. 

I don't have a strong preference here, so I am open to your opinions on this. Just thinking out loud I want to add just the following couple of comments:
1. We can also have methods for computing stuff on sections that are not methods of sections class, but methods that require as input a section, for instance `compute_crack_width(section, ...)`
2. Having at disposal the material permits also to do on-the-fly changes on the material and see immediately the results on the section response as observed during tests by me and @mortenengen. 
3. On this last point we should anyhow account that running the code for a section is easy. And maybe running two cells with two set of parameters permits comparisons, while updating material prevents easy comparisons since you "overwrite" the old result. Maybe we can improve the API adding methods for geometries permitting to update the constitutive law passing the newly updated material?
4. If we want to include material objects in the section class, I would prevent the user to pass a constitutive law, since that if we feel the necessity of having materials for computations, it means that constitutive laws are not enough.

Let me know you thoughts on this.

**Comment by mortenengen** [comment]
These are all good points, and I am glad that they are brought up now 😃 

tl;dr: I suggest that we explore implementing separate utility functions for concrete, but I am not yet sure how to handle the reference to material classes. More below.

I think there are some important considerations that you also bring up:
- Storing the material class in the section can be useful, but really requires that we only accept a material class and not either a material or a constitutive law. Being able to provide just a constitutive law is in my opinion really strong, since it opens for a lot of flexibility. Therefore, I think we should stick to the current implementation, i.e. accepting either a material or a constitutive law, and storing only the constitutive law on the section.
- We have planned to create an element class, which as you indicate @DanielGMorenaFhecor, is above the section classes in the hierarchy. And as @talledodiego indicates, a method like the suggested `.wk` could belong there. However, I am not really sure if it actually belongs there either, because then we might end in a situation where we have a too close connection between the design code equations and the higher level classes, i.e. mixing concerns. I think a good alternative is to define separate utility functions for concrete, as @talledodiego indicates, perhaps even design code specific ones, where the section is one of the arguments. But this does not solve the original issue in this thread: how to access properties from the material class. One way forward with such utility functions could be to provide both a section and a material as arguments, but then we are doubling up on the material definition, the procedure becomes ambiguous, and it is unclear how to treat a situation where you have different concretes in the same section.
- Finally, we might consider an option where a constitutive law stores a reference to the material class that initialized it. If it was not initialized by a material class, it could store `None`. If in addition we implement separate utility functions for concrete as indicated above, we could raise meaningful exceptions if these functions are called with sections that contain constitutive laws without any reference to materials.

What are your thoughts?
